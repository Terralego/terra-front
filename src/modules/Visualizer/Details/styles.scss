@import '../../../utils/variables.scss';

.view-details {
  position: absolute;
  top: 0.5rem;
  bottom: .5rem;
  right: 0;
  width: 25rem;
  display: flex;
  flex-direction: column;
  transform: translate3d(100%, 0, 0);
  transition: transform .2s ease-in, opacity .2s ease-in;
  z-index: 2;
  background: #fff;
  margin: 0;
  opacity: 0;
  padding: 0 2em;

  &--hidden {
    display: none;
  }

  &--visible {
    transform: translate3d(-.5rem, 0, 0);
    opacity: .9;
    box-shadow: 0 0 10px rgba(0,0,0,.2);
  }

  &--withCarousel {
    padding: 0;
  }

  &__close-button {
    position: absolute;
    top: .5rem;
    right: .5rem;
    cursor: pointer;
  }

  &__wrapper {
    display: flex;
    overflow-x: hidden;
    overflow-y: auto;
    height: 100%;
  }

  &__content {
    padding: .5rem 0;
    flex: auto;
    min-width: 0;
  }

  &__button {
    position: relative;
    transition: transform 0.2s ease-in;
    height: 100%;
    width: 1.75rem;
    flex-shrink: 0;
    button {
      position: absolute;
      top: 12.5%;
      right: 0;
      left: 0;
      transform: translateY(-50%) scale(1.5);
    } 
  }

  .details {
    display: flex;
    flex-direction: column;
    justify-content: center;

    &__loading {
      position: absolute;
      top: 50%;
      left: 50%;
      transform: translate(-50%);
    }
    
    &__title {
      text-align: center;
      font-size: 1.6rem;
      font-weight: 300;
    }

    &__group {
      margin-top: .5rem;
    }

    &__list {
      padding: 0;
      margin: 0;
      display: table;
      
      &__amenity {
        display: flex;
        flex-direction: row;
        justify-content: space-between;
        padding: 0;

        &__img {
          width: 2.25rem;
          height:2.25rem;
        }
      }
    }

    &__times {
      display: flex;
      flex-direction: row;
      justify-content: space-around;
      padding: 0;
    }

    &__subtitle {
      color: #106ba3;
      opacity: .65;
      font-weight: normal;
    }

    &__column {
      display: table-row;

      &-label {
        display: table-cell;
        padding-bottom: .5rem;
        white-space: nowrap;
        opacity: .65;
      }

      &-value {
        display: table-cell;
        padding-bottom: 10px;
        overflow-wrap: anywhere; // Firefox
        word-break: break-word; // Chrome
        // If no support, it will fallback with a scrollbar

        &-time {
          display: flex;
          flex-direction: column;
          justify-content: center;
          list-style: none;
        }
      }
    }

    &__link {
      margin-top: 3.5rem;
      text-align: center;

      a {
        background: #fff;
        padding: .8rem;
        border: 1px solid rgba(10, 6, 77, .35);
      }
    }
  }
}

@media(min-width: $breakpoints-tablet) and (max-width: $breakpoints-desktop) and (orientation: landscape) {
  .view-details {
    width: 16.5rem;
  
    &__close-button {
      top: .25rem;
      right: .25rem;
    }
  
    &__wrapper {
      display: flex;
      overflow-x: hidden;
      overflow-y: auto;
      height: 100%;
    }
  
    &__content {
      flex: auto;
      min-width: 0;
    }
  
    .details {
      padding: .5rem / $scaleRatio;

      &__picto {
        width: 4.25rem;
        margin: auto;
      }
  
      &__title {
        font-size: 1.6rem/$scaleRatio;
        margin: .75rem 0;
      }
  
      &__group {
        margin-top: .5rem/$scaleRatio;
      }
      &__group:first-child {
        margin-top: 0;
      }
  
      &__list {
        padding: 0;
        margin: 0;
        display: flex;
        flex-direction: column;
        
        &__amenity {
          display: flex;
          flex-direction: row;
          justify-content: space-between;
          padding: 0;
  
          &__img {
            width: 2.25rem/$scaleRatio;
            height: 2.25rem/$scaleRatio;
          }
        }
      }
  
      &__times {
        display: flex;
        flex-direction: row;
        justify-content: space-around;
        padding: 0;
      }
  
      &__column {
        display: flex;
        flex-direction: column;
  
        &-label {
          display: flex;
          padding-bottom: .5rem/$scaleRatio;
        }
  
        &-value {
          display: flex;
          padding-bottom: 1.2rem;
          
          &-time {
            display: flex;
            flex-direction: column;
            justify-content: center;
            list-style: none;
          }
        }
      }
    }
    &__button {
      height: 100%;
      width: .75rem;
      flex-shrink: 0;
      &--prev {
        transform: translateX(-100%);
      }
      &--next {
        transform:translateX(100%);
      }
      &--active {
        transform: translateX(0)
      }

      button {
        top: 15%;
        margin: 0;
        min-width: .75rem;
        min-height: .75rem;
        padding: 0;
        outline: none;
      } 
    }
  }
}

@media(min-width: $breakpoints-phone) and (max-width: $breakpoints-tablet) and (orientation: portrait) {
  .view-details {
    width: 16.5rem;
  
    &__close-button {
      top: .25rem;
      right: .25rem;
    }
  
    &__wrapper {
      display: flex;
      overflow-x: hidden;
      overflow-y: auto;
      height: 100%;
    }
  
    &__content {
      flex: auto;
      min-width: 0;
    }
  
    .details {
      padding: .5rem / $scaleRatio;

      &__picto {
        width: 4.25rem;
        margin: auto;
      }
  
      &__title {
        font-size: 1.6rem/$scaleRatio;
        margin: .75rem 0;
      }
  
      &__group {
        margin-top: .5rem/$scaleRatio;
      }
      &__group:first-child {
        margin-top: 0;
      }
  
      &__list {
        padding: 0;
        margin: 0;
        display: flex;
        flex-direction: column;
        
        &__amenity {
          display: flex;
          flex-direction: row;
          justify-content: space-between;
          padding: 0;
  
          &__img {
            width: 2.25rem/$scaleRatio;
            height: 2.25rem/$scaleRatio;
          }
        }
      }
  
      &__times {
        display: flex;
        flex-direction: row;
        justify-content: space-around;
        padding: 0;
      }
  
      &__column {
        display: flex;
        flex-direction: column;
  
        &-label {
          display: flex;
          padding-bottom: .5rem/$scaleRatio;
        }
  
        &-value {
          display: flex;
          padding-bottom: 1.2rem;
          
          &-time {
            display: flex;
            flex-direction: column;
            justify-content: center;
            list-style: none;
          }
        }
      }
    }
    &__button {
      height: 100%;
      width: .75rem;
      flex-shrink: 0;
      &--prev {
        transform: translateX(-100%);
      }
      &--next {
        transform:translateX(100%);
      }
      &--active {
        transform: translateX(0)
      }

      button {
        top: 15%;
        margin: 0;
        min-width: .75rem;
        min-height: .75rem;
        padding: 0;
        outline: none;
      } 
    }
  }
}
