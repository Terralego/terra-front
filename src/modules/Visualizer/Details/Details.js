import React from 'react';
import PropTypes from 'prop-types';
import classnames from 'classnames';
import { Button } from '@blueprintjs/core';

import FeatureProperties from '../../Map/FeatureProperties';
import Template from '../../Template';
import './styles.scss';

class Details extends React.Component {
  state = {
    index: -1,
    hidden: true,
  }

  static getDerivedStateFromProps ({ feature, interaction }) {
    /**
     * Keep a copy of props to have a nice animation with previous content
     */
    if (feature) {
      return { feature, interaction };
    }
    return null;
  }

  componentDidMount () {
    const {
      feature: { properties: { _id } = {} } = {},
    } = this.props;
    this.updateIndex(_id);
  }

  componentDidUpdate ({
    visible: prevVisible,
    feature: { properties: { _id: prevId } = {} } = {},
  }) {
    const {
      visible,
      feature: { properties: { _id } = {} } = {},
    } = this.props;

    if (prevId !== _id) {
      this.updateIndex(_id);
    }

    if (visible !== prevVisible) {
      this.switchVisibility(visible);
    }
  }

  getIndexFeature = index => {
    const { features } = this.props;
    const featuresLength = features.length;
    if (index <= -1) {
      return featuresLength - 1;
    }
    if (index === featuresLength) {
      return 0;
    }
    return index;
  }

  handleChange = direction => {
    const { onChange } = this.props;
    const { index: prevIndex = 0 } = this.state;
    const index = this.getIndexFeature(prevIndex + direction);
    onChange(index);
    return this.setState({ index });
  }

  switchVisibility (visible) {
    if (visible) {
      clearTimeout(this.hideTimeout);
      this.setState({
        hidden: false,
      }, () => this.setState({
        visible: true,
      }));
    } else {
      this.setState({
        visible: false,
      });
      this.hideTimeout = setTimeout(() => this.setState({ hidden: true }), 200);
    }
  }

  updateIndex (id) {
    const { features = [] } = this.props;
    if (features.length) {
      this.setState({ index: features.findIndex(({ _id }) => _id === id) });
    }
  }

  render () {
    const {
      features,
      onClose = () => null,
      isTableActive,
    } = this.props;
    const {
      feature: { properties } = {},
      interaction: { template, fetchProperties = {} } = {},
      index, hidden, visible,
    } = this.state;
    const isCarrousel = features.length > 1;
    const featureToDisplay = features.length > 0 && index !== -1 ? features[index] : properties;

    return (
      <div
        className={classnames(
          {
            'view-details': true,
            'bp3-light': true,
            'view-details--hidden': hidden,
            'view-details--visible': visible,
            'view-details--withTable': isTableActive,
            'view-details--withCarousel': isCarrousel,
          },
        )}
      >
        <div className="view-details__close">
          <Button
            type="button"
            className="view-details__close-button"
            onClick={onClose}
            icon="cross"
            minimal
          />
        </div>
        {featureToDisplay && (
          <div className="view-details__wrapper">
            {isCarrousel && (
              <div
                className={classnames(
                  'view-details__button',
                  'view-details__button--prev',
                )}
              >
                <Button
                  type="button"
                  onClick={() => this.handleChange(-1)}
                  icon="chevron-left"
                  minimal
                />
              </div>
            )}
            <div className="view-details__content">
              <FeatureProperties
                {...fetchProperties}
                properties={featureToDisplay}
              >
                {({ properties: newProperties, ...staticProperties }) => (
                  <Template
                    template={template}
                    {...staticProperties}
                    {...newProperties}
                  />
                )}
              </FeatureProperties>
            </div>
            {isCarrousel && (
              <div
                className={classnames(
                  'view-details__button',
                  'view-details__button--next',
                )}
              >
                <Button
                  type="button"
                  onClick={() => this.handleChange(1)}
                  icon="chevron-right"
                  minimal
                />
              </div>
            )}
          </div>
        )}
      </div>
    );
  }
}

Details.propTypes = {
  onChange: PropTypes.func,
};

Details.defaultProps = {
  onChange () {},
};

export default Details;
