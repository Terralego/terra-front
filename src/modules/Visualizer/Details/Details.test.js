import React from 'react';
import renderer from 'react-test-renderer';
import { shallow } from 'enzyme';
import { Button } from '@blueprintjs/core';
import Details from './Details';
import FeatureProperties from '../../Map/FeatureProperties';

jest.useFakeTimers();

jest.mock('@blueprintjs/core', () => ({
  Button () {
    return <p>Button</p>;
  },
}));

jest.mock('../../Map/FeatureProperties', () =>
  ({ children }) => children({ properties: {} }));

it('should render correctly with content', () => {
  const tree = renderer.create((
    <Details
      visible
      interaction={{
        template:
`# Hello world
 This is {{foo}}`,
      }}
    />
  )).toJSON();
  expect(tree).toMatchSnapshot();
});

it('should render correctly without content', () => {
  const tree = renderer.create((
    <Details
      interaction={{
        template:
  `# Hello world
  This is {{foo}}`,
      }}
    />
  )).toJSON();
  expect(tree).toMatchSnapshot();
});

it('should close', () => {
  const onClose = jest.fn();
  const wrapper = shallow((
    <Details
      onClose={onClose}
    />
  ));
  wrapper.find('Button').props().onClick();
  expect(onClose).toHaveBeenCalled();
});

it('should do nothing', () => {
  const onClose = jest.fn();
  const wrapper = shallow((
    <Details features={[]} />
  ));
  wrapper.find('Button').props().onClick();
  expect(onClose).not.toHaveBeenCalled();
});

it('should get derived state from props', () => {
  const feature = {};
  const interaction = {};
  expect(Details.getDerivedStateFromProps({ feature, interaction }))
    .toEqual({ feature, interaction });
  expect(Details.getDerivedStateFromProps({ }))
    .toBeNull();
});

it('should update component', () => {
  const instance = new Details({});
  instance.updateIndex = jest.fn();
  instance.switchVisibility = jest.fn();

  instance.props.visible = true;
  instance.componentDidUpdate({
    visible: false,
  });
  expect(instance.switchVisibility).toHaveBeenCalledWith(true);
  instance.switchVisibility.mockClear();

  instance.componentDidUpdate({
    visible: true,
  });
  expect(instance.switchVisibility).not.toHaveBeenCalled();
  instance.switchVisibility.mockClear();

  instance.props.feature = { properties: { _id: 1 } };
  instance.componentDidUpdate({
    feature: {},
  });
  expect(instance.updateIndex).toHaveBeenCalledWith(1);
  instance.updateIndex.mockClear();
});

it('should get index feature', () => {
  const features = [{}, {}, {}];
  const instance = new Details({ features });
  expect(instance.getIndexFeature(0)).toBe(0);
  expect(instance.getIndexFeature(1)).toBe(1);
  expect(instance.getIndexFeature(2)).toBe(2);
  expect(instance.getIndexFeature(3)).toBe(0);
  expect(instance.getIndexFeature(-1)).toBe(2);
});

it('should switch visibility on', () => {
  const instance = new Details();
  let callback;
  instance.setState = jest.fn((state, cb) => { callback = cb; });
  instance.switchVisibility(true);
  expect(instance.setState).toHaveBeenCalledWith({
    hidden: false,
  }, callback);
  expect(clearTimeout).toHaveBeenCalledTimes(1);
  instance.setState.mockClear();
  callback();
  expect(instance.setState).toHaveBeenCalledWith({
    visible: true,
  });
});

it('should switch visibility off', () => {
  const instance = new Details();
  instance.setState = jest.fn();
  instance.switchVisibility(false);
  expect(instance.setState).toHaveBeenCalledWith({
    visible: false,
  });
  expect(setTimeout).toHaveBeenCalledTimes(1);
  instance.setState.mockClear();
  jest.runAllTimers();
  expect(instance.setState).toHaveBeenCalledWith({ hidden: true });
});

it('should update index', () => {
  const features = [{
    _id: 1,
  }, {
    _id: 2,
  }, {
    _id: 3,
  }];
  const instance = new Details({});
  instance.setState = jest.fn();
  instance.updateIndex(1);
  expect(instance.setState).not.toHaveBeenCalled();

  instance.props.features = features;
  instance.updateIndex(1);
  expect(instance.setState).toHaveBeenCalledWith({ index: 0 });
  instance.setState.mockClear();

  instance.updateIndex(2);
  expect(instance.setState).toHaveBeenCalledWith({ index: 1 });
  instance.setState.mockClear();

  instance.updateIndex(3);
  expect(instance.setState).toHaveBeenCalledWith({ index: 2 });
  instance.setState.mockClear();

  instance.updateIndex(4);
  expect(instance.setState).toHaveBeenCalledWith({ index: -1 });
});

it('should handle change', () => {
  const features = [{}, {}, {}];
  const onChange = jest.fn();
  const instance = new Details({ features, onChange });
  instance.setState = jest.fn();

  instance.handleChange(1);
  expect(onChange).toHaveBeenCalledWith(0);
  expect(instance.setState).toHaveBeenCalledWith({ index: 0 });
  onChange.mockClear();

  instance.state.index = 0;
  instance.handleChange(1);
  expect(onChange).toHaveBeenCalledWith(1);
  expect(instance.setState).toHaveBeenCalledWith({ index: 1 });
  onChange.mockClear();

  instance.state.index = 1;
  instance.handleChange(1);
  expect(onChange).toHaveBeenCalledWith(2);
  expect(instance.setState).toHaveBeenCalledWith({ index: 2 });
  onChange.mockClear();

  instance.state.index = 2;
  instance.handleChange(1);
  expect(onChange).toHaveBeenCalledWith(0);
  expect(instance.setState).toHaveBeenCalledWith({ index: 0 });
  onChange.mockClear();

  instance.state.index = undefined;
  instance.handleChange(-1);
  expect(onChange).toHaveBeenCalledWith(2);
  expect(instance.setState).toHaveBeenCalledWith({ index: 2 });
  onChange.mockClear();

  instance.state.index = 2;
  instance.handleChange(-1);
  expect(onChange).toHaveBeenCalledWith(1);
  expect(instance.setState).toHaveBeenCalledWith({ index: 1 });
  onChange.mockClear();

  instance.state.index = 1;
  instance.handleChange(-1);
  expect(onChange).toHaveBeenCalledWith(0);
  expect(instance.setState).toHaveBeenCalledWith({ index: 0 });
  onChange.mockClear();

  instance.state.index = 0;
  instance.handleChange(-1);
  expect(onChange).toHaveBeenCalledWith(2);
  expect(instance.setState).toHaveBeenCalledWith({ index: 2 });
  onChange.mockClear();
});

it('should display a feature', () => {
  const wrapper = shallow(
    <Details
      features={[
        {}, {}, {},
      ]}
    />,
  );
  const instance = wrapper.instance();
  instance.setState({ index: 0 });
  expect(wrapper.find('.view-details__wrapper').length).toBe(1);
  const { onClick: down } = wrapper.find(Button).get(1).props;
  const { onClick: up } = wrapper.find(Button).get(2).props;
  instance.handleChange = jest.fn();

  down();
  expect(instance.handleChange).toHaveBeenCalledWith(-1);

  up();
  expect(instance.handleChange).toHaveBeenCalledWith(1);

  const { children } = wrapper.find(FeatureProperties).props();
  expect(children({})).toMatchSnapshot();
});
