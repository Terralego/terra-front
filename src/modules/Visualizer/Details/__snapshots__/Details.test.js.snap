// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should display a feature 1`] = `null`;

exports[`should render correctly with content 1`] = `
<div
  className="view-details bp3-light view-details--hidden"
>
  <div
    className="view-details__close"
  >
    <p>
      Button
    </p>
  </div>
</div>
`;

exports[`should render correctly without content 1`] = `
<div
  className="view-details bp3-light view-details--hidden"
>
  <div
    className="view-details__close"
  >
    <p>
      Button
    </p>
  </div>
</div>
`;
